core/commands/ls.go

```diff
@@ -71,14 +81,20 @@ The JSON output contains type information.
 			return
 		}
 
+		rtype, rf, _ := req.Option("type").String()
+		if rf && rtype != resolveAlways && rtype != resolveLocal && rtype != resolveNever {
+			res.SetError(errInvalidResolveType, cmdkit.ErrClient)
+			return
+		}
+
 		resolve, _, err := req.Option("resolve-type").Bool()
 		if err != nil {
 			res.SetError(err, cmdkit.ErrNormal)
 			return
 		}
```

Comment:
I would change
```go
resolve, _, err := req.Option("resolve-type").Bool()
if err != nil {
	res.SetError(err, cmdkit.ErrNormal)
	return
}
```
to
```go
{
	resolve, defined, err := req.Option("resolve-type").Bool()
	if err != nil {
		res.SetError(err, cmdkit.ErrNormal)
		return
	}
	if rf && defined {
		res.SetError(fmt.Errorf("both 'resove-type' and 'resolve' can not be specified", cmdkit.ErrNormal))
		return
	}
	if resolve {
		rtype = resolveAlways
	} else {
		rtype = resolveNever
	}
}
```
(but not I did not test nor even try to compile this code)

This (1) makes sure both `resolve-type` and `resolve` are not both specified. (2) sets rtype to the correct value if `resolve` is specified and (3) prevents the use of the resolve variable elsewhere by putting the entire thing in its own scope.

Reply:
1) If we set default value for at least one of this options, it will become difficult to check for conflicts. For example, in your variant, if we set default value for `--resolve`, we will always have an error, once we have set `--resolve-type` (if I've understood everything right).

2) I had no default value for the new option, because i wanted to keep interface backwards compatible. If this is not a problem, may be it is better to make `--resolve-type` a string value as @Stebalien had proposed in earlier comment?

Will submit a new request soon.

Reply:
@frrist that is not how default values work, the second value returned is only true if the user sets the value, if it is false than the first value returned is the default value.  All options have a default, it is wither false, the empty string, or 0, using WithDefault just makes the default explicit.

There will only be a conflict if both are specified on the command line.  If this is the case both `defined` and `rf` will be true.

Reply:
@kevina oh, then I'm wrong, thought it works the other way, will change then.

Reply:
@kevina i've checked this again and i think second return value is true every time default value is set (both in PreRun and Run).
We invoke FillDefaults() in Parse() https://github.com/ipfs/go-ipfs-cmds/blob/master/cli/parse.go#L23
Which in turn can modify original request with default values: https://github.com/ipfs/go-ipfs-cmds/blob/master/request.go#L158

Reply:
@fyrchik that does not sound right and might be a bug if it is true.  I would test this.  Also have a look at the command for the add command.

But maybe I am misunderstand something.  @whyrusleeping @Stebalien any insight?

Also @Stebalien in 4909c5a56d256e0b999d11b065227b94576609ce you removed the `WithDefault(0)` for `cidVersionOptionName`
```
-		cmdkit.IntOption(cidVersionOptionName, "Cid version. Non-zero value will change default of 'raw-leaves' to true. (experimental)").WithDefault(0),
+		cmdkit.IntOption(cidVersionOptionName, "CID version. Defaults to 0 unless an option that depends on CIDv1 is passed. (experimental)"),
```
why?  Does it relate to what @fyrchik is describing.

Reply:
Also see #4911.